sql/item_float_interchange.cc

```diff
@@ -0,0 +1,109 @@
+/* Copyright (c) 2018 MariaDB Foundation
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#include "mariadb.h"
+#include "item_float_interchange.h"
+
+
+#if float32_exists
+///////////////////////////////////////////////////////////////////////////
+longlong Item_func_interchange_float32_to_uint32::val_int()
+{
+  return (longlong)val_uint();
+}
+
+ulonglong Item_func_interchange_float32_to_uint32::val_uint()
+{
+  float32 f;
+  uint32 u;
+  DBUG_ASSERT(fixed);
+  DBUG_ASSERT(sizeof(float32) == sizeof(uint32));
+
+  f= args[0]->val_real();
+  u= float32_to_uint32(f);
+  null_value= args[0]->null_value;
+
+  return u;
+}
+
+///////////////////////////////////////////////////////////////////////////
+
+double Item_func_interchange_uint32_to_float32::val_real()
+{
+  float32 f;
+  uint32 u;
+  DBUG_ASSERT(fixed);
+  DBUG_ASSERT(sizeof(float32) == sizeof(uint32));
+
+  // If a Item_ulonglong_func arrives, we should switch to that,
+  // and use args[0]->val_uint();
+  // Alexander Barkov informs me:
+  // Currently val_uint() just reinterprets the result of val_int()
+  // from signed to unsigned.
+  // But this will be changed soon:
+  // it will cut negative values to 0 (with a warning).
+  u= (uint32)args[0]->val_int();
+  f= uint32_to_float32(u);
+  null_value= isfinite(f) ? args[0]->null_value : 1;
```

Comment:
I think the assignment above does something not really useful for negative values as well as huge positive values.
I think it's better to cut out-of-range values to the valid range 0..UINT_MAX32.
Something like this should do:

u= (uint32) args[0]->val_uint_from_val_int(UINT_MAX32);

Reply:
@abarkov, thanks for the review and the idea.

If input is an out-of-range value, then perhaps an error is best. I will experiment with some tests.